parser/parser.go

```diff
@@ -59,60 +57,51 @@ type (
 
 func RunParser(reader io.Reader) Parser {
 	p := &pcmParser{
-		source:          reader,
-		keysReady:       make(chan struct{}),
-		keysInfoMutex:   new(sync.RWMutex),
-		streamInfoReady: make(chan struct{}),
-		streamInfoMutex: new(sync.RWMutex),
+		source: reader,
+		mutex:  new(sync.RWMutex),
 	}
 	go p.run()
-	return p
-}
 
-func RunStreamedParser(reader io.Reader, chanLen int) (Parser, <-chan ValuesOrError) {
-	p := &pcmParser{
-		source:        reader,
-		sink:          make(chan ValuesOrError, chanLen),
-		keysReady:     make(chan struct{}),
-		keysInfoMutex: new(sync.RWMutex),
-	}
-	go p.run()
+	sig := make(chan struct{})
+	go func() {
+		for {
+			if p.isReady {
+				close(sig)
+				return
+			}
+		}
+	}()
 
-	return p, p.sink
+	for {
+		select {
+		case <-time.After(time.Second * 2):
+			log.Fatal("timed out starting parser")
+		case <-sig:
+			return p
+		}
+	}
 }
 
 func (p *pcmParser) GetKeys(ctx context.Context) ([]Key, error) {
-	var sig chan struct{}
-	withRLock(p.keysInfoMutex, func() {
-		sig = p.keysReady
+	sig := make(chan struct{})
+	go withRLock(p.mutex, func() {
+		close(sig)
 	})
+
 	for {
 		select {
 		case <-ctx.Done():
 			return nil, ctx.Err()
 		case <-sig:
-			var keys []Key
-			withRLock(p.keysInfoMutex, func() {
-				keys = p.keys
-			})
-			return keys, nil
+			return p.keys, nil
```

Comment:
There is also possible data race, because even before `select` mutex is unlocked, reading goroutine can start writing keys.